File: python/Algorithms/Sorting/merge_sort.py
from typing import List
from math import floor
from Algorithms.Sorting.common import CompareFunction, Sortable


def merge_sort(input_list: List[Sortable],
               comparator: CompareFunction) -> List[Sortable]:
    """"
    Sorts the input list, using the provided comparator
    to decide if two elements are in the correct order.
    Returns a copy of the list sorted, makes no changes to input list.
    This function delegates to the recursive form, initialising the left and right pointers.
    """

    # Empty list, or single item list, just return it
    if len(input_list) < 2:
        return list(input_list)

    return merge_sort_recurse(input_list, comparator, 0, len(input_list) - 1)


def merge_sort_recurse(input_list: List[Sortable],
                       comparator: CompareFunction,
                       left_pointer: int,
                       right_pointer: int) -> List[Sortable]:
    """"
    Recursive form of the merge sort that requires the left and right pointers.
    """
    # Exit condition, the list is a single item
    if left_pointer == right_pointer:
        return [input_list[left_pointer]]

    # Calculate the mid point
    middle: int = floor((left_pointer + right_pointer) / 2)

    # Recurse sort both halves to yield two lists to merge
    first_half: List[Sortable] = merge_sort_recurse(
        input_list, comparator, left_pointer, middle)
    second_half: List[Sortable] = merge_sort_recurse(
        input_list, comparator, middle + 1, right_pointer)

    # Merge the two halves into a single sorted list
    output_list: List[Sortable] = []
    while (len(first_half) > 0) and (len(second_half) > 0):
        # Comparator returns -1 if the first item is 'larger than' the first
        comparison: int = comparator(first_half[0], second_half[0])
        if comparison < 0:
            output_list.append(second_half.pop(0))
        else:
            output_list.append(first_half.pop(0))

    # Push any stragglers from whichever list has items remaining
    output_list.extend(first_half)
    output_list.extend(second_half)

    return output_list

File: lib/src/algorithms/sort/mergeSort.ts
import { SplitList } from "../../types";
import {
  emptyObserver,
  anyComparator,
  ROOT_RECURSION_KEY,
} from "../../common";
import { SortUtility } from "./types";

function mergeSortR<T>(
  inputList: T[],
  utilities: SortUtility<T>,
  leftPointer?: number,
  rightPointer?: number,
  parentKey: number = ROOT_RECURSION_KEY,
  levelAdjust: number = ROOT_RECURSION_KEY / 2
): T[] {

  // Have we reached the bottom of our recursion? This is the exit condition!
  if (leftPointer === rightPointer) {
    return [inputList[leftPointer]];
  }

  const {
    compare = anyComparator,
    observe = emptyObserver,
    split = emptyObserver,
    join = emptyObserver,
  } = utilities;

  // Calculate the mid point
  const middle = Math.floor((leftPointer + rightPointer) / 2);

  const listA: SplitList<T> = {
    key: (parentKey - levelAdjust).toString(10),
    data: inputList.slice(leftPointer, middle + 1),
  };
  const listB: SplitList<T> = {
    key: (parentKey + levelAdjust).toString(10),
    data: inputList.slice(middle + 1, rightPointer + 1),
  };
  observe("Recursing", inputList, { leftPointer, rightPointer, middle });
  split(parentKey.toString(10), listA, listB);

  // Recurse sort both halves to yield the two lists to merge
  const firstHalf = mergeSortR(
    inputList,
    utilities,
    leftPointer,
    middle,
    parentKey - levelAdjust,
    levelAdjust / 2
  );
  const secondHalf = mergeSortR(
    inputList,
    utilities,
    middle + 1,
    rightPointer,
    parentKey + levelAdjust,
    levelAdjust / 2
  );

  // Merge the two halves into a single sorted list
  const outputList = [];
  let firstPtr = 0;
  let secondPtr = 0;
  while (firstPtr < firstHalf.length && secondPtr < secondHalf.length) {
    // Comparator returns +ve if the second item is larger than first
    if (
      compare(firstHalf[firstPtr], secondHalf[secondPtr], {
        aIndex: leftPointer + firstPtr,
        bIndex: middle + secondPtr,
      }) > 0
    ) {
      outputList.push(secondHalf[secondPtr]);
      secondPtr += 1;
    } else {
      outputList.push(firstHalf[firstPtr]);
      firstPtr += 1;
    }
  }

  // Push any stragglers from whichever list has items remaining
  firstHalf.filter((_, i) => i >= firstPtr).forEach((i) => outputList.push(i));
  secondHalf
    .filter((_, i) => i >= secondPtr)
    .forEach((i) => outputList.push(i));
  join(listA, listB, outputList);

  return outputList;
}

function mergeSort<T>(inputList: T[],
  utilities: SortUtility<T>): T[] {

  // Is it worth sorting?
  if (inputList.length < 2) {
    return inputList;
  }

  return mergeSortR(inputList, utilities, 0, inputList.length - 1)
}

export default mergeSort;
